File: leetcode/array/953. Verifying an Alien Dictionary/js/main.js
// Time O(N^2)
// Space O(N)
const isAlienSorted = (words, order) => {
  let map = new Map();
  let i = 0;

  for (let ch of order) {
    map.set(ch, i++);
  }

  let nonSortedWords = [...words];
  words.sort(compare);

  for (let i = 0; i < nonSortedWords.length; i++) {
    if (nonSortedWords[i] != words[i]) {
      return false;
    }
  }

  return true;

  function compare(a, b) {
    for (let i = 0; i < Math.min(a.length, b.length); i++) {
      if (a[i] == b[i]) {
        continue;
      }

      return map.get(a[i]) - map.get(b[i]);
    }

    return a.length - b.length;
  }
};

File: leetcode/array/953. Verifying an Alien Dictionary/py/main.py
class Solution:
    def isAlienSorted(self, words, order):
        mapping = {ch: i for i, ch in enumerate(order)}

        return words == sorted(words, key=lambda s: [mapping[ch] for ch in s])

    def isAlienSorted_II(self, words, order):
        mapping = {ch: i for i, ch in enumerate(order)}
        n = len(words)

        for i in range(n - 1):
            j = 0
            word1 = words[i]
            word2 = words[i + 1]
            min_lenght = min(len(word1), len(word2))

            while j < min_lenght:
                if word1[j] != word2[j]:
                    if mapping[word1[j]] > mapping[word2[j]]:
                        return False
                    break
                j += 1

            if j == min_lenght and word1[j - 1] == word2[
                    j - 1] and len(word1) > len(word2):
                return False

        return True
